### PR TITLE
Implement chain sync manager

### PR DESCRIPTION
**Pull request by whyrusleeping**
The chain sync manager is responsible both for scheduling concurrent sync operations, and also for selecting the correct heads to sync to initially during bootstrap. 

I've still got a bit to do on this, but I wanted to open it up for some early review

### CODE REVIEW

File: chain/sync_manager.go [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+
+func (sm *SyncManager) syncScheduler() {
+	var syncQueue syncBucketSet
+	var activeSyncTips syncBucketSet
+
+	var nextSyncTarget *syncTargetBucket
+	var workerChan chan *types.TipSet
+
+	for {
+		select {
```
Comment (magik6k) [inline chain/sync_manager.go]: I'd put those cases in separate functions

Comment (whyrusleeping) [reply chain/sync_manager.go]: yeah... i was thinking i might have to do that. Will do it now

File: chain/sync.go [hunk 2]
```diff
@@ -758,9 +752,20 @@ func (syncer *Syncer) verifyBlsAggregate(ctx context.Context, sig types.Signatur
 	return nil
 }
 
+const syncStateKey = "syncStateKey"
```
Comment (Kubuxu) [inline chain/sync.go]: The context value key should be a custom non-exported type.
so:
```
type syncStateKey struct{}

v := ctx.Value(syncStateKey{})
```

**Comment by whyrusleeping** [comment]
Yeah, its coupled on purpose. I initially was thinking of just making it part of the Syncer, but i wanted to separate it just enough to make testing it easier.

File: chain/sync_manager.go [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+	}
+	return nil
+}
+
+func (sbs *syncBucketSet) Heaviest() *types.TipSet {
+	// TODO: should also consider factoring in number of peers represented by each bucket here
+	var bestTs *types.TipSet
+	for _, b := range sbs.buckets {
+		bhts := b.heaviestTipSet()
+		if bestTs == nil || bhts.ParentWeight().GreaterThan(bestTs.ParentWeight()) {
```
Comment (magik6k) [inline chain/sync_manager.go]: This should probably not be the parent ts weight (but the one from `ChainStore.Weight`)

File: chain/sync_manager.go [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+}
+
+func (stb *syncTargetBucket) heaviestTipSet() *types.TipSet {
+	if stb == nil {
+		return nil
+	}
+
+	var best *types.TipSet
+	for _, ts := range stb.tips {
+		if best == nil || ts.ParentWeight().GreaterThan(best.ParentWeight()) {
```
Comment (magik6k) [inline chain/sync_manager.go]: ditto

File: chain/sync_manager.go [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+				sm.nextSyncTarget = relbucket
+				sm.workerChan = sm.syncTargets
+			} else {
+				sm.syncQueue.buckets = append(sm.syncQueue.buckets, relbucket)
+			}
+		} else {
+			// TODO: this is the case where we try to sync a chain, and
+			// fail, and we have more blocks on top of that chain that
+			// have come in since.  The question is, should we try to
+			// sync these? or just drop them?
```
Comment (magik6k) [inline chain/sync_manager.go]: If we can retry later I'd say dropping them is fine

File: chain/syncstate.go [hunk 6]
```diff
@@ -32,12 +32,20 @@ type SyncerState struct {
 }
 
 func (ss *SyncerState) SetStage(v api.SyncStateStage) {
+	if ss == nil {
+		return
```
Comment (magik6k) [inline chain/syncstate.go]: log.Warn in these?

File: cli/sync.go [hunk 7]
```diff
@@ -65,19 +68,38 @@ var syncWaitCmd = &cli.Command{
...
 				target = ss.Target.Cids()
 			}
 
-			fmt.Printf("\r\x1b[2KTarget: %s\tState: %s\tHeight: %d", target, chain.SyncStageString(ss.Stage), ss.Height)
-			if ss.Stage == api.StageSyncComplete {
-				fmt.Println("\nDone")
+			fmt.Printf("\r\x1b[2KWorker %d: Target: %s\tState: %s\tHeight: %d", working, target, chain.SyncStageString(ss.Stage), ss.Height)
+
+			if time.Now().Unix()-int64(head.MinTimestamp()) < build.BlockDelay {
+				fmt.Println("Done!")
```
Comment (magik6k) [inline cli/sync.go]: That print above doesn't print the new line (and shouldn't)

File: chain/sync_manager.go (hunk 3 above)
Comment (whyrusleeping) [reply chain/sync_manager.go]: You can't get the weight of a tipset without evaluating it, and in order to evaluate it, you have to sync it, and this is the logic that is deciding what to sync

File: chain/sync_manager.go (hunk 4 above)
Comment (whyrusleeping) [reply chain/sync_manager.go]: ditto

File: chain/sync_manager.go (hunk 5 above)
Comment (whyrusleeping) [reply chain/sync_manager.go]: Yeah, we can definitely retry later. I think I agree, dropping is fine

File: chain/syncstate.go (hunk 6 above)
Comment (whyrusleeping) [reply chain/syncstate.go]: ehhhh, the point here is for things like tests to be able to run sync tests without having to inject a syncerstate tracker

File: cli/sync.go (hunk 7 above)
Comment (whyrusleeping) [inline cli/sync.go]:
```suggestion
				fmt.Println("\nDone!")
```

**Pull request merged**